include/alpaka/block/shared/st/BlockSharedMemStMember.hpp

```diff
@@ -118,18 +43,25 @@ namespace alpaka
             //-----------------------------------------------------------------------------
             static auto declareVar(BlockSharedMemStMember<TDataAlignBytes> const& smem) -> T&
             {
-                smem.template alloc<T>();
-                return smem.template getLatestVar<T>();
+                auto* data = smem.template getVarPtr<T>(TuniqueId);
+
+                if(!data)
+                {
+                    smem.template alloc<T>(TuniqueId);
+                    data = smem.template getLatestVarPtr<T>();
+                }
+                ALPAKA_ASSERT(data != nullptr);
+                return std::ref(*data);
             }
         };
         //#############################################################################
         template<std::size_t TDataAlignBytes>
         struct FreeSharedVars<BlockSharedMemStMember<TDataAlignBytes>>
         {
             //-----------------------------------------------------------------------------
-            static auto freeVars(BlockSharedMemStMember<TDataAlignBytes> const& mem) -> void
+            static auto freeVars(BlockSharedMemStMember<TDataAlignBytes> const&) -> void
             {
-                mem.free();
+                // shared memory block data will be reused
             }
         };
```

Comment:
From a quick look at the alpaka code base, it seems that was the last use of `FreeSharedVars` that actually does something. So we can remove the entire trait and all specializations. Mind that `AccOacc` has an implementation of `FreeSharedVars`, but it delegates again `freeSharedVars`. So either it recurses indefinitely or it does nothing ;)

Reply:
Because from the concept it is correct to have a `free` method to flush or delete the memory. I already deleted it but reactivated the code to keep the door open that someone can implement the shared memory with other allocators.

Reply:
I think this is a case of YAGNI.

Reply:
It was required when the CPU accelerator were allocating shared memory dynamically on the heap. It is obsolete at the moment.

Reply:
Go ahead, leave it in if you want. I would drop it, decreasing the legacy.

Reply:
I have actually no strong opinion on dropping or leaving (slightly leaning towards dropping, actually), except, that this change would belong into a separate PR.